### PR TITLE
WIP: Initial flatbuffer rr_graph.

### PR DESCRIPTION
**Pull request by litghost**

### CODE REVIEW

File: vpr/rr_graph.fbs [hunk 1]
```diff
@@ -0,0 +1,176 @@
+namespace vtr.vpr;
+
+table ChannelList {
```
Comment (mithro) [inline vpr/rr_graph.fbs:3]: I'm not sure we need ChannelList / Channels - they seem like they could be automatically calculated on the fly?

**Comment by mcmasterg** [comment]
Have you done any profiling? What sort of memory footprint change did this have?

File: vpr/rr_graph.fbs [hunk 2]
```diff
@@ -0,0 +1,176 @@
...
+  x:int;
+  y:int;
+  block_type_id:int;
+  width_offset:int;
+  height_offset:int;
+}
+
+enum NodeType: byte {
+  INVALID_NODE_TYPE,
+  CHANX,
```
Comment (mithro) [inline vpr/rr_graph.fbs:93]: Probably really only need `CHAN`, it's a `CHANX` if xlow == xhigh and `CHANY` if ylow == yhigh.

File: vpr/rr_graph.fbs [hunk 3]
```diff
@@ -0,0 +1,176 @@
+namespace vtr.vpr;
```
Comment (mithro) [inline vpr/rr_graph.fbs:1]: This was my attempt at an rr_graph.fbs file;
```
// example IDL file

namespace VtR;

attribute "priority";

enum RRNodeType : byte {
  CHANX = 1,
  CHANY,
  SOURCE,
  SINK,
  OPIN,
  IPIN,
}

enum RRNodeSide : byte {
  LEFT = 1,
  RIGHT,
  TOP,
  BOTTOM,
}

struct Coord {
  x:uint;
  y:uint;
}

struct RRNodeChannel {
  id:uint (key, required);
  start:Coord;
  end:Coord;
  track:uint;
}
/// If start.x == end.x, then CHANX
/// If start.y == end.y, then CHANY

/// If pin_class.OUTPUT then SOURCE
/// If pin_class.INPUT or pin_class.CLOCK then SINK
struct RRNodeEndpoint {
  id:uint (key, required);
  pos:Coord;
  pin_class: ??;
}

struct RRNodePin {
  id:uint (key, required);
  pos:Coord;
  pin_index: ??;
  side:RRNodeSide;
}

union RRNode { RRNodeChannel, RRNodeEndpoint, RRNodePin }

struct RREdge {
  src:??;
  sink:??;
  switch:??;
}

table Monster {
  pos:Vec3;
  mana:short = 150;
  hp:short = 100;
  name:string;
  friendly:bool = false (deprecated, priority: 1);
  inventory:[ubyte];
  color:Color = Blue;
  test:Any;
}

file_identifier "VtRR";
file_extension "rr_graph.fb";

root_type Monster;
```

Comment (mithro) [inline vpr/rr_graph.fbs:1]: This file should probably be somewhere other than the root directory?

File: vpr/src/route/rr_graph_reader.cpp [hunk 4]
```diff
@@ -15,6 +15,9 @@
```
Comment (mithro) [inline vpr/src/route/rr_graph_reader.cpp:15]: This should probably be two files now? One for the XML version and one for the flatbuffer version?

**Review by mithro — COMMENTED** [review]
Some initial random comments.

**Comment by mithro** [comment]
I think you have done a direct translation of the xml format into flatbuffer format? The plan would be to then evolve the flatbuffer and xml formats in tandem?

That seems like an initially a good approach compared to my attempt at doing a different format.

**Comment by litghost** [comment]
Next step is to get rrgraph patching using flatbuffers.  

Because of how flatbuffers is implemented, that in memory cost is roughly equal to the disk size.  As an initial preview using VPR only, outputting the virtual rrgraph flatbuffer for the ice40 8k part takes ~4 seconds and the disk size of the rrgraph is 46 MB.  The XML takes ~9 seconds and the disk size is 140 MB.  Note that the XML representation is not actually modelled, VPR directly writes the XML.

Reading the virtual 8k table with VPR:

flatbuffer:	

command: $(which time) -v $(pwd)/../vpr/vpr ../../symbiflow-arch-defs-cmake/build/ice40/devices/top-routing-virt/arch.merged.xml --device hx8k-bg121 ../../symbiflow-arch-defs-cmake/common/wire.eblif --read_rr_graph test.rrgraph --place --pack --route_chan_width 100

Results:
	User time (seconds): 0.75
	System time (seconds): 0.22
	Elapsed (wall clock) time (h:mm:ss or m:ss): 0:00.86
	Maximum resident set size (kbytes): 129324


xml:
command:$(which time) -v $(pwd)/../vpr/vpr ../../symbiflow-arch-defs-cmake/build/ice40/devices/top-routing-virt/arch.merged.xml --device hx8k-bg121 ../../symbiflow-arch-defs-cmake/common/wire.eblif --read_rr_graph test.xml --place --pack --route_chan_width 100

Results:
	User time (seconds): 1.65
	System time (seconds): 0.46
	Elapsed (wall clock) time (h:mm:ss or m:ss): 0:02.03
	Maximum resident set size (kbytes): 710680


So that's a pretty big win in both resident memory and CPU.

**Comment by litghost** [comment]
Tim: Ya, its a straight translation.  I was spinning my wheels a little trying to come up with a more perfect format, so I figured this both improves the state of vpr, and gives me time to process what is actually happening in the rr graph format.

**Comment by mithro** [comment]
FYI - Looking at the stats you gave, the XML versions are not problematic? 
```
Elapsed (wall clock) time (h:mm:ss or m:ss): 0:02.03
Maximum resident set size (kbytes): 710680
```

Wall clock of 2 seconds and memory size of 710mb seems pretty reasonable.

BTW Do you want to just try adding gzip xml support? Should be a pretty quick thing to do.

**Comment by litghost** [comment]
"Wall clock of 2 seconds and memory size of 710mb seems pretty reasonable." - That's for an 8k part and a virtual routing list.  An ice40 8k has 1561942 edges in the virtual graph and 3267202 in the real graph.

For referencing, the states for VPR on the REAL rr_graph.xml is as follows:

xml:
	User time (seconds): 6.42
	System time (seconds): 1.34
	Elapsed (wall clock) time (h:mm:ss or m:ss): 0:07.70
	Maximum resident set size (kbytes): 2244320

flatbuffer:
	User time (seconds): 4.42
	System time (seconds): 0.63
	Elapsed (wall clock) time (h:mm:ss or m:ss): 0:04.92
	Maximum resident set size (kbytes): 791332

So we are looking at now ~2 GB to hold the XML in memory, and we are still only talking about it 8k part.  The flatbuffer version is still consume 800 MB, but the flatbuffer is only 151 MB, so the rest of the ~650 MB is vpr itself.  Assuming that the VPR consumes the same RAM for XML and flatbuffer (which is a reasonable assumption), then it took ~1350 MB to model a 350 MB XML file, versus ~151 MB.  That's a factor of ten in resident memory usage.

**Comment by litghost** [comment]
Re: gzip xml -> I'm not super concerned about the on-disk size.  Resident memory and parse times are more relevant.  The problem with the rrgraph xml is really the fact that there is ~4x inflation between the disk size and the resident memory size, combined with ~2-3x file size cost.

**Comment by mithro** [comment]
You might find that using a gzip file ends up being faster as it takes less time to read from disk and decompress than reading the decompressed file.

**Comment by mithro** [comment]
@litghost 

I don't think you should need to hold the whole XML file memory either? You should be able to generate and write to disk as you go along and read from disk and throw away as parsing? (IE Often called the ["sax" model of XML processing](https://en.wikipedia.org/wiki/Simple_API_for_XML).)

**Comment by mithro** [comment]
@litghost You might find this github issue interesting -> https://github.com/zeux/pugixml/issues/135 - specially the example from [@halx99](https://github.com/zeux/pugixml/issues/135#issuecomment-321952406) and [@silviucpp](https://github.com/zeux/pugixml/issues/135#issuecomment-325306639). 

The results of the [xmldrv3](https://github.com/halx99/xmldrv3) example seem interesting...

**Pull request closed without merging**